File: js/maybe-monad.js
var maybeMonad = new function() {
  var wrap = function(value) {
    return function() { return value }
  };

  var bind = function(mv, f) {
    return function() {
      if (mv()) {
        return f(mv())();
      } else {
        return mv();
      }
    }
  };
  return {
    wrap: wrap,
    bind: bind
  }
}();

exports.maybeMonad = maybeMonad;

File: clojure/test/monads_blog/test/maybe_monad.clj
(ns monads-blog.test.maybe-monad
  (:use [monads-blog.maybe-monad])
  (:use [monads-blog.test.monadic-laws])
  (:use [clojure.test]))

; dummy example functions to use for tests
(defn more[value]
    (wrap (str "more " value)))

(defn much[value]
    (wrap (str "more " value)))

; actual tests
(deftest bind-test
  (testing "when calling the resulting monadic value we get nil if the original value was nil"
    (is (= ((bind (wrap nil) more)) nil))))

(deftest monadic-laws
  (testing "the maybe monad follows the monadic laws"
     (is (first-monadic-law wrap bind more "candy"))
     ; interesting case: should functions test for nil as well?
     ; (is (first-monadic-law wrap bind more nil))
     (is (second-monadic-law wrap bind "candy"))
     (is (second-monadic-law wrap bind nil))
     (is (third-monadic-law wrap bind much more "candy"))))
